### PR TITLE
Fight Sequence

### PR DESCRIPTION
**Pull request by MLHumphries**
Added checks to see if players can attack, calculates attack damage (based on weapon damage vs. armor), and reduces player/enemy health and stamina.

### CODE REVIEW

File: dunwoody_disaster/CharacterFactory.py [hunk 1]
```diff
@@ -10,12 +10,14 @@ def __init__(self):
         self.maxHealth = 0
         self.curMagic = 0
         self.maxMagic = 0
-        self.curMechanical = 0
-        self.maxMechanical = 0
+        self.curStamina = 0
+        self.maxStamina = 0
```
Comment (chanson02) [inline dunwoody_disaster/CharacterFactory.py:14]: I wonder how many problems this is going to cause...

File: FightSequence.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+from dunwoody_disaster import CharacterFactory, Item
+from dunwoody_disaster.views import fightScreen
+
+
+class FightSequence:
+    def __init__(self, player, enemy):
+        self.player = player
+        self.enemy = enemy
+
+    def Fight(self, playerWeapon, enemyWeapon):
```
Comment (chanson02) [inline FightSequence.py]:
```suggestion
    def Fight(self, playerWeapon: Item.Weapon, enemyWeapon: Item.Weapon) -> tuple:
```

File: FightSequence.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            fightScreen.UpdateMeters(self.player, self.player.meters)
+        if canEnemyAttack:
+            enemyDamage = self.CalculateDamage(enemyWeapon, self.enemy, self.player)
+            self.enemy.curStamina -= enemyWeapon.staminaCost
+            self.player.health -= enemyDamage
+            fightScreen.UpdateMeters(self.enemy, self.enemy.meters)
+
+        return self.player, self.enemy
+
+    def CanAttack(self, player, attack) -> bool:
```
Comment (chanson02) [inline FightSequence.py]:
```suggestion
    def CanAttack(self, player: CharacterFactory.Character, attack: Item.Weapon) -> bool:
```

File: FightSequence.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        Checks to see if character has enough stamina or magic to attack with.
```
Comment (chanson02) [inline FightSequence.py:42]: Good work on the docstrings! Keep it up!

File: FightSequence.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        if (
+            player.curStamina - attack.staminaCost >= 0
+            and player.curMagic >= attack.magicCost
+        ):
+            return True
+        else:
+            return False
```
Comment (chanson02) [inline FightSequence.py]:
```suggestion
        return player.curStamina - attack.staminaCost >= 0 and player.curMagic >= attack.magicCost
```

File: FightSequence.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    def CalculateDamage(playerAttack: Item.Weapon, targetDefense: Item.Armor):
+        """
+        Deals damage based on playerAttack vs the targets defensive item.
+        """
+        attackDamage = playerAttack.damage - targetDefense.armorVal
+
+        return attackDamage
```
Comment (chanson02) [inline FightSequence.py:55]:
```suggestion
        return playerAttack.damage - targetDefense.armorVal
```

File: FightSequence.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            player.curStamina - attack.staminaCost >= 0
+            and player.curMagic >= attack.magicCost
+        ):
+            return True
+        else:
+            return False
+
+    def CalculateDamage(playerAttack: Item.Weapon, targetDefense: Item.Armor):
+        """
+        Deals damage based on playerAttack vs the targets defensive item.
```
Comment (chanson02) [inline FightSequence.py]: Does this function actually `deal` the damage, or does it just calculate how much damage the weapon _WOULD DO_ against the opponent

File: FightSequence.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+from dunwoody_disaster import CharacterFactory, Item
+from dunwoody_disaster.views import fightScreen
+
+
+class FightSequence:
+    def __init__(self, player, enemy):
```
Comment (chanson02) [inline FightSequence.py]:
```suggestion
    def __init__(self, player: CharacterFactory.Character, enemy: CharacterFactory.Character):
```

File: FightSequence.py (hunk 2 above)
Comment (chanson02) [reply FightSequence.py]: There is some way to make this like `-> tuple[CharacterFactory.Character]` but I don't remember exactly how.

File: dunwoody_disaster/CharacterFactory.py [hunk 9]
```diff
@@ -10,12 +10,14 @@ def __init__(self):
...
 
-        # What are these? --Cooper
-        self.defense = 0
-        self.magicDefense = 0
+        self.meters = {
+            "health": self.maxHealth,
+            "magic": self.maxMagic,
+            "stamina": self.maxStamina,
+        }
```
Comment (chanson02) [inline dunwoody_disaster/CharacterFactory.py:20]: What is the point of this? I have something similar but a little different on my latest branch. Don't want to cause any conflicts.

File: dunwoody_disaster/Item.py [hunk 10]
```diff
@@ -28,9 +28,12 @@ def __str__(self) -> str:
 
 
 class Weapon(Item):
-    def __init__(self, name):
+    def __init__(self, name, damage, magicCost, staminaCost):
```
Comment (chanson02) [inline dunwoody_disaster/Item.py:31]:
```suggestion
    def __init__(self, name: str, damage: int, magicCost: int, staminaCost: int):
```

File: dunwoody_disaster/Item.py [hunk 11]
```diff
@@ -40,6 +43,7 @@ def __init__(self, name):
 
 
 class Armor(Item):
-    def __init__(self, name):
+    def __init__(self, name, armorVal):
```
Comment (chanson02) [inline dunwoody_disaster/Item.py]:
```suggestion
    def __init__(self, name: str, armorVal: int):
```

**Review by chanson02 — CHANGES_REQUESTED** [review]
Typehints and a couple simplifications

File: dunwoody_disaster/CharacterFactory.py (hunk 9 above)
Comment (MLHumphries) [reply dunwoody_disaster/CharacterFactory.py:20]: We had talked about putting the meters on the player, so I did that.

File: dunwoody_disaster/views/fightScreen.py [hunk 12]
```diff
@@ -329,13 +329,13 @@ def UpdateMeters(self, player: Character, meters: dict):
             magicPercentage = (curMagic / maxMagic) * 100
         meters["magic"].setPercentage(magicPercentage)
 
-        maxMechanical = player.maxMechanical
-        curMechanical = player.curMechanical
-        if curMechanical == 0:
-            mechanicalPercentage = 0
+        maxStamina = player.maxMechanical
+        curStamina = player.curMechanical
```
Comment (chanson02) [inline dunwoody_disaster/views/fightScreen.py]: These attributes no longer exist.

File: dunwoody_disaster/views/fightScreen.py [hunk 13]
```diff
@@ -329,13 +329,13 @@ def UpdateMeters(self, player: Character, meters: dict):
...
+        if curStamina == 0:
+            staminaPercentage = 0
         else:
-            mechanicalPercentage = (curMechanical / maxMechanical) * 100
-        meters["mechanical"].setPercentage(mechanicalPercentage)
+            staminaPercentage = (curStamina / maxStamina) * 100
+        meters["mechanical"].setPercentage(staminaPercentage)
```
Comment (chanson02) [inline dunwoody_disaster/views/fightScreen.py]: This key no longer exists.

File: dunwoody_disaster/CharacterFactory.py [hunk 14]
```diff
@@ -125,12 +127,10 @@ def createCharacter(name: str, classType: str) -> Character:
...
         character.maxHealth = data["health"]
         character.maxMagic = data["magic"]
-        character.maxMechanical = data["mechanical"]
+        character.maxStamina = data["mechanical"]
         character.curHealth = data["health"]
         character.curMagic = data["magic"]
-        character.curMechanical = data["mechanical"]
-        character.defense = data["defense"]
-        character.magicDefense = data["magicDefense"]
+        character.curStamina = data["mechanical"]
```
Comment (chanson02) [inline dunwoody_disaster/CharacterFactory.py]: We should phase out `data["mechanical"]`. The `CharacterFactory` is going to need some tweaks.

**Review by chanson02 — CHANGES_REQUESTED** [review]
`Mechanical` -> `Stamina` bugs

File: dunwoody_disaster/CharacterFactory.py (hunk 14 above)
Comment (chanson02) [reply dunwoody_disaster/CharacterFactory.py]: I see  that you changed it to use `data['stamina']`, but you also need to change it on `CharacterFactory.class_types`

**Review by chanson02 — APPROVED** [review]
Appears to be working.

**Review by instMeyFTW — APPROVED** [review]

**Pull request merged**